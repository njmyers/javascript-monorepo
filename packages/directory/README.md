# Directory

Reads all the files within a directory. Returns absolute or relative paths, mime types and file contents. Filter your files based on file extensions and recursively (or not) read the whole directory. Like `'fs'` but with some extras for sure!

## Usage

```js
import directory from '@njmyers/directory';

const files = directory('../path/to/folder/or/file', { ...Options })

const Options = {
  /** boolean option to return mime information for all files: default false */
  mime: boolean,
  /** boolean option to return absolute paths for all files default: true */
  absolute: boolean,
  /** boolean option to search the directory recursively default: false */
  recursive: boolean,
  /** boolean option to read the files and return the contents as a string default: false */
  read: boolean,
  /** string or array of file types to filter (include) default: false */
  filter: string | Array<string>,
};

/** Returns an Array of FileObjects */
const FileObject = {
  /** the path of the file */
  path: string,
  /** whether it was included (will always be true) */
  include: boolean,
  /** the mime information */
  mime: {
    /** the content type ie text/css */
    contentType: string,
    /** the file extension with no leading period */
    extension: string,
  },
  /** the file contents */
  file: string,
};
```