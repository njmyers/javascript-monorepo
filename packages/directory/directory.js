import fs from 'fs';
import path from 'path';
import lodash from 'lodash';

const isFile = (file) => fs.statSync(file).isFile();

const readDirectoryKeys = (dir) => {
  return isFile(dir)
    ? dir
    : lodash.flatten(
        fs.readdirSync(dir).map((name) => {
          const filePath = `${dir}/${name}`;
          return isFile(filePath) ? filePath : readDirectoryKeys(filePath);
        })
      );
};

export default readDirectoryKeys;
