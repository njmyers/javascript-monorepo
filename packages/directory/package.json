{
  "name": "@njmyers/directory",
  "version": "0.0.12",
  "description": "Reads all the files within a directory.",
  "license": "ISC",
  "author": "Nick Myers <nickmyers111@gmail.com>",
  "main": "build/directory.cjs.js",
  "module": "build/directory.esm.js",
  "scripts": {
    "prepublishOnly": "yarn run build",
    "build": "run-s rollup",
    "rollup": "BABEL_ENV=production rollup -c",
    "test": "jest --watchAll --no-cache"
  },
  "dependencies": {
    "@babel/runtime": "^7.1.2",
    "mime-types": "^2.1.19",
    "smalldash": "^1.1.17"
  },
  "devDependencies": {
    "@njmyers/babel-runtime-files": "^0.0.3",
    "jest": "^23.4.1",
    "npm-run-all": "^4.1.3",
    "shx": "^0.3.2"
  },
  "gitHead": "42bfcd1db1c84f88dc8e242f8856e78ffb754477"
}
