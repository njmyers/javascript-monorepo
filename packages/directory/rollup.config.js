import { es6, cjs } from '@njmyers/rollup-config';
import pkg from './package.json';

const external = ['fs', 'path'];

const cjsConfig = cjs(pkg);
const es6Config = es6(pkg);

export default [
  {
    input: 'src/index.ts',
    ...cjsConfig,
    external: [...cjsConfig.external, ...external],
  },
  {
    input: 'src/index.ts',
    ...cjsConfig,
    external: [...cjsConfig.external, ...external],
  },
];
