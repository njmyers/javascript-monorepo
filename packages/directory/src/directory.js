// @flow
import * as fs from 'fs';
import * as path from 'path';
import { pipe } from 'smalldash';
import * as mime from 'mime-types';
// types
import type { Mime, FileObject, Options } from './types';

/** is it a file or a dir? */
const isFile = (file) => fs.statSync(file).isFile();

/**
 * Reads a directory and returns an array of their paths.
 * If you put in relative paths you will recieve relative paths.
 * If you input absolute paths you will receive absolute paths.
 * @param {string} dir        the path to read
 * @param {Options} options   the options object
 * @return {array}            the files inside the folder
 */
function readDirectory(dir: string, options: Options = {}): Array<string> {
  return isFile(dir)
    ? [dir]
    : []
        .concat(
          ...fs.readdirSync(dir).map((name) => {
            const filePath = `${dir}/${name}`;
            return options.recursive
              ? isFile(filePath)
                ? filePath
                : readDirectory(filePath, options)
              : isFile(filePath)
                ? filePath
                : '';
          })
        )
        .filter((string) => string);
}

/**
 * Helper function returns an absolute path
 * @param  {[type]} relative [description]
 * @return {[type]}          [description]
 */
function pathify(relative: string): string {
  return path.resolve(relative);
}

/**
 * Turns the path into an object with property 'path'
 * @param  {string} path    an absolute path string
 * @return {FileObject}     a file object
 */
function objectify(path: string): FileObject {
  return { path };
}

/**
 * Adds mime information to the file information
 * @param  {FileObject} obj a file object with the required path property
 * @return {FileObject}     a file object with mime types added in
 */
function mimeify(obj: FileObject): FileObject {
  return {
    ...obj,
    mime: {
      contentType: mime.lookup(obj.path),
      extension: path.extname(obj.path).slice(1),
    },
  };
}

/**
 * Reads the file and adds it to the object
 * @param  {FileObject} obj a file object with the required path property
 * @return {FileObject}     a file object with the file contents added in
 */
function read(obj: FileObject): FileObject {
  return {
    ...obj,
    file: fs.readFileSync(obj.path, 'utf8').toString(),
  };
}

/** default options */
const O = {
  absolute: true,
  recursive: false,
  mime: false,
  read: false,
};

/**
 * Looks up files with options including a return of relative or absolute paths
 * Recursively look up the directory
 * Add mime contentType and extension
 * @param  {string} dir       the relative or absolute path to read
 * @param  {Options} options  the options object
 * @return {array}            an array of file objects containing your information
 */
function directory(dir: string, options: Options = O): Array<FileObject> {
  const list = readDirectory(dir, options);

  const pipeline = []
    .concat(options.absolute || options.read ? pathify : [])
    .concat(options.mime || options.read ? objectify : [])
    .concat(options.mime ? mimeify : [])
    .concat(options.read ? read : []);

  return pipeline.length > 0 ? list.map(pipe(...pipeline)) : list;
}

export default directory;

export { readDirectory, objectify, pathify, mimeify, read };
