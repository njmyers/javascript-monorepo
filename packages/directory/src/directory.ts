import { pipe } from 'smalldash';

import {
  filterify,
  mimeify,
  pathify,
  objectify,
  readDirectorySync,
  readSync,
} from './utils';

// types
import { FileObject, Options } from './types';

/** default options */
const defO = {
  absolute: true,
  recursive: false,
  mime: false,
  read: false,
  filter: false,
};

/**
 * Looks up files with options including a return of relative or absolute paths
 * Recursively look up the directory
 * Add mime contentType and extension
 * @param  {string} dir       the relative or absolute path to read
 * @param  {Options} options  the options object
 * @return {array}            an array of file objects containing your information
 */
function directory(dir: string, options: Options = defO): Array<FileObject> {
  const list = readDirectorySync(dir, options);

  const pipeline = []
    .concat(options.absolute || options.read ? pathify : [])
    .concat(options.mime || options.read || options.filter ? objectify : [])
    .concat(options.mime || options.filter ? mimeify : [])
    .concat(options.filter ? filterify(options.filter) : [])
    .concat(options.read ? readSync : []);

  return pipeline.length > 0
    ? list
        // map through our pipeline
        .map(pipe(...pipeline))
        // finally apply the filtering for our safe results array
        // if we are in "only paths" mode then we won't have an object
        // if we have an object we default to include unless we add a filter
        .filter(obj => obj.include || typeof obj === 'string')
    : list;
}

export default directory;
