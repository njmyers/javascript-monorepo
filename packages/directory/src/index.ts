export { default } from './directory';
export { default as directory } from './directory';
export { default as directoryAsync } from './directory-async';
