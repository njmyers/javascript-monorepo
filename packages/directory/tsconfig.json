{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "outDir": "build",
    "rootDir": ".",
    "baseUrl": ".",
    "paths": {
      "smalldash/*": ["../smalldash/src/*"],
      "~/*": ["src/*"]
    }
  }
}
