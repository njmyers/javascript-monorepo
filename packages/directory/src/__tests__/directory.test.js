import fs from 'fs';
import path from 'path';
import directory, { readDirectory, read } from '../directory';

const homedir = require('os').homedir();

const testDirectory = __dirname;

/**
 * These tests can be maddening.
 * Since we run jest from the repo root process.cwd() is the repo root.
 * Remember that and try not to run jest from any other directory.
 * Otherwise you may literally lose your sanity.
 */

describe('it reads the directory - unit test for readDirectory', () => {
  test('sanity test - if this fails run jest from the monorepo root', () => {
    expect(__dirname).toBe(
      path.resolve(process.cwd(), 'packages/directory/src/__tests__')
    );
  });

  test('it reads a file defined locally', () => {
    expect(readDirectory('package.json')).toMatchObject(['package.json']);
  });

  /** This ends the local section from here on let's deal with absolute paths */

  test('it reads a file defined absolutely', () => {
    expect(
      readDirectory(path.resolve(__dirname, 'helpers/test.txt'), {
        recursive: true,
      })
    ).toMatchObject([path.resolve(__dirname, 'helpers/test.txt')]);
  });

  test('it reads a folder defined absolutely of depth = 1', () => {
    expect(
      readDirectory(path.resolve(__dirname, 'helpers/folder'))
    ).toMatchObject([
      `${__dirname}/helpers/folder/file.txt`,
      `${__dirname}/helpers/folder/otherfile.txt`,
    ]);
  });

  test('it reads a folder defined absolutely of depth > 1', () => {
    expect(
      readDirectory(path.resolve(__dirname, 'helpers/recursive'), {
        recursive: true,
      })
    ).toMatchObject([
      `${__dirname}/helpers/recursive/file.txt`,
      `${__dirname}/helpers/recursive/folder/file.txt`,
      `${__dirname}/helpers/recursive/folder/otherfile.js`,
      `${__dirname}/helpers/recursive/folder/thing/file.txt`,
    ]);
  });
});

describe('it reads the directory and uses options', () => {
  /** testing with local paths */
  test('it returns absolute paths automatically', () => {
    const pkg = directory('package.json');

    expect(directory('package.json')).toMatchObject([
      path.resolve(process.cwd(), 'package.json'),
    ]);
  });

  /** testing with absolute paths */
  test('it reads a folder defined absolutely of depth > 1', () => {
    expect(
      directory(path.resolve(__dirname, 'helpers/recursive'), {
        recursive: true,
      })
    ).toMatchObject([
      `${__dirname}/helpers/recursive/file.txt`,
      `${__dirname}/helpers/recursive/folder/file.txt`,
      `${__dirname}/helpers/recursive/folder/otherfile.js`,
      `${__dirname}/helpers/recursive/folder/thing/file.txt`,
    ]);
  });

  test('it reads a file and adds mime information', () => {
    expect(
      directory(path.resolve(__dirname, 'helpers/test.txt'), {
        mime: true,
      })
    ).toMatchObject([
      {
        path: `${__dirname}/helpers/test.txt`,
        include: true,
        mime: {
          contentType: 'text/plain',
          extension: 'txt',
        },
      },
    ]);
  });

  test('it reads a directory and adds mime information', () => {
    expect(
      directory(path.resolve(__dirname, 'helpers/recursive'), {
        mime: true,
        recursive: true,
      })
    ).toMatchObject([
      {
        mime: { contentType: 'text/plain', extension: 'txt' },
        path: `${__dirname}/helpers/recursive/file.txt`,
        include: true,
      },
      {
        mime: { contentType: 'text/plain', extension: 'txt' },
        path: `${__dirname}/helpers/recursive/folder/file.txt`,
        include: true,
      },
      {
        mime: { contentType: 'application/javascript', extension: 'js' },
        path: `${__dirname}/helpers/recursive/folder/otherfile.js`,
        include: true,
      },
      {
        mime: { contentType: 'text/plain', extension: 'txt' },
        path: `${__dirname}/helpers/recursive/folder/thing/file.txt`,
        include: true,
      },
    ]);
  });

  test('it reads a directory and adds file contents', () => {
    expect(
      directory(path.resolve(__dirname, 'helpers/recursive'), {
        recursive: true,
        read: true,
      })
    ).toMatchObject([
      {
        file: '',
        include: true,
        path: `${__dirname}/helpers/recursive/file.txt`,
      },
      {
        file: '',
        include: true,
        path: `${__dirname}/helpers/recursive/folder/file.txt`,
      },
      {
        file: `/** File */\n\nconst thing = 'arbitrary javascript';\n`,
        include: true,
        path: `${__dirname}/helpers/recursive/folder/otherfile.js`,
      },
      {
        file: '',
        include: true,
        path: `${__dirname}/helpers/recursive/folder/thing/file.txt`,
      },
    ]);
  });

  test('it reads a directory and filters by file extension', () => {
    expect(
      directory(path.resolve(__dirname, 'helpers/recursive'), {
        recursive: true,
        read: true,
        filter: 'js',
      })
    ).toMatchObject([
      {
        file: `/** File */\n\nconst thing = 'arbitrary javascript';\n`,
        include: true,
        path: `${__dirname}/helpers/recursive/folder/otherfile.js`,
      },
    ]);
  });

  test('it reads a directory and filters by array of file extensions', () => {
    expect(
      directory(path.resolve(__dirname, 'helpers/recursive'), {
        recursive: true,
        read: true,
        filter: ['js'],
      })
    ).toMatchObject([
      {
        file: `/** File */\n\nconst thing = 'arbitrary javascript';\n`,
        include: true,
        path: `${__dirname}/helpers/recursive/folder/otherfile.js`,
      },
    ]);

    expect(
      directory(path.resolve(__dirname, 'helpers/recursive'), {
        recursive: true,
        read: true,
        filter: ['js', 'txt'],
      })
    ).toMatchObject([
      {
        file: '',
        include: true,
        path: `${__dirname}/helpers/recursive/file.txt`,
      },
      {
        file: '',
        include: true,
        path: `${__dirname}/helpers/recursive/folder/file.txt`,
      },
      {
        file: `/** File */\n\nconst thing = 'arbitrary javascript';\n`,
        include: true,
        path: `${__dirname}/helpers/recursive/folder/otherfile.js`,
      },
      {
        file: '',
        include: true,
        path: `${__dirname}/helpers/recursive/folder/thing/file.txt`,
      },
    ]);
  });

  test('it reads a directory and filters by malformed file extension', () => {
    const result = [
      {
        file: `/** File */\n\nconst thing = 'arbitrary javascript';\n`,
        include: true,
        path: `${__dirname}/helpers/recursive/folder/otherfile.js`,
      },
    ];

    expect(
      directory(path.resolve(__dirname, 'helpers/recursive'), {
        recursive: true,
        read: true,
        filter: '.js',
      })
    ).toMatchObject(result);

    expect(
      directory(path.resolve(__dirname, 'helpers/recursive'), {
        recursive: true,
        read: true,
        filter: ['.js'],
      })
    ).toMatchObject(result);
  });
});
