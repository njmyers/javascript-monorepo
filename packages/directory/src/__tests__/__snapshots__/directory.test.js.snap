// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it reads the directory and uses options it reads a directory and adds file contents 1`] = `
Array [
  Object {
    "file": "",
    "path": "/home/developer/Documents/monorepo/packages/directory/src/__tests__/helpers/recursive/file.txt",
  },
  Object {
    "file": "",
    "path": "/home/developer/Documents/monorepo/packages/directory/src/__tests__/helpers/recursive/folder/file.txt",
  },
  Object {
    "file": "/** File */

const thing = 'arbitrary javascript';
",
    "path": "/home/developer/Documents/monorepo/packages/directory/src/__tests__/helpers/recursive/folder/otherfile.js",
  },
  Object {
    "file": "",
    "path": "/home/developer/Documents/monorepo/packages/directory/src/__tests__/helpers/recursive/folder/thing/file.txt",
  },
]
`;
